### PR TITLE
[DAP-5348] Pass in keyspace-level vschema vindex configuration info to newVindexFunc() so that vindexes can access config info for other vindexes defined in the same vschema.

### PR DESCRIPTION
**Pull request by swu-etsy**
## Context

A [previous commit](https://github.com/etsy/vitess/commit/a9fdeb467d6088c87d52c0d5f251ba7fa138b7be) added a new Multisharded vindex type. The Multisharded vindex receives two record columns -- a type_id, which corresponds to a shardifier type and its corresponding hybrid vindex, and the identifier id. The Multisharded vindex uses the hybrid vindex corresponding to the type_id to resolve the identifier id to a key range. 

However, the approach of the previous commit relied on an incorrect assumption that vindexes would be created in the order that they are defined in the vschema.json file. Specifically, the previous approach stored [a pointer to each hybrid vindex](https://github.com/etsy/vitess/blob/a9fdeb467d6088c87d52c0d5f251ba7fa138b7be/go/vt/vtgate/vindexes/hybrid.go#L89) as the vindex was created, and [validated](https://github.com/etsy/vitess/blob/a9fdeb467d6088c87d52c0d5f251ba7fa138b7be/go/vt/vtgate/vindexes/multisharded.go#L64) that all hybrid vindexes that the Multisharded vindex depended on (I'll refer to them as the Multisharded vindex's "subvindexes") existed at the time the Multisharded vindex was created. As it turns out, the order in which vindexes are created is non-deterministic. That's because Vitess [unmarshals the vschema.json into a map](https://github.com/etsy/vitess/blob/a9fdeb467d6088c87d52c0d5f251ba7fa138b7be/go/vt/proto/vschema/vschema.pb.go#L156), which is [unordered in Golang](https://go.dev/blog/maps#:~:text=When%20iterating%20over%20a%20map%20with%20a%20range%20loop%2C%20the%20iteration%20order%20is%20not%20specified%20and%20is%20not%20guaranteed%20to%20be%20the%20same%20from%20one%20iteration%20to%20the%20next.).

As a result, the `No hybrid vindex [...] has been defined` error is sometimes thrown when a vschema containing a Multisharded vindex is loaded.

## This PR's Changes

This PR eliminates the Multisharded vindex's dependency on vschema creation order. It does so by [passing in a vschemapb.Keyspace.Vindex object](https://github.com/etsy/vitess/blob/a9fdeb467d6088c87d52c0d5f251ba7fa138b7be/go/vt/vtgate/vindexes/vschema.go#L253), containing keyspace-level vindex vschema config info, to each vindex's [NewVindexFunc](https://github.com/etsy/vitess/blob/93ff149cc43d36958e7f5d6504c434b98b9ad00a/go/vt/vtgate/vindexes/vindex.go#L189) when each vindex defined in vschema.json is created.

The new `ksVindexesSchemaInfo` parameter allows the `NewMultiSharded` function to access vschema config data [necessary to create the subvindexes it depends on](https://github.com/etsy/vitess/blob/93ff149cc43d36958e7f5d6504c434b98b9ad00a/go/vt/vtgate/vindexes/multisharded.go#L74) when a Multisharded vindex is created. 

Currently, this new parameter is only used by the Multisharded vindex type. This PR contains changes across many files because each vindex's `NewVindexFunc` is updated to support an additional parameter, and callsites for `CreateVindex` and implementations of `NewVindexFunc` must be updated to pass in the new `ksVindexesSchemaInfo` argument. In most cases, `nil` is passed in for non-Multisharded vindexes.

## Risks

1. Compared to the previous approach, the current approach is more memory intensive, because Hybrid vindex instances will be created multiple times: first, when they are created as a result of being independently defined in the vschema.json, and again when they are created by any Multisharded vindex(es). However, there are precedents for this pattern in the [hybrid vindex](https://github.com/etsy/vitess/blob/a9fdeb467d6088c87d52c0d5f251ba7fa138b7be/go/vt/vtgate/vindexes/hybrid.go#L77) and the [multicol vindex](https://github.com/etsy/vitess/blob/a9fdeb467d6088c87d52c0d5f251ba7fa138b7be/go/vt/vtgate/vindexes/multicol.go#L176), both of which create multiple hash vindex instances with identical configurations.

2. It only passes a non-nil value for the `ksVindexesSchemaInfo` argument when `CreateVindex` is called within the BuildTables method. If `CreateVindex` is ever called with `NewMultiSharded` elsewhere, the function would error. I searched the codebase for references to `CreateVindex`, and didn't find any other callsites where it would be called with `NewMultiSharded`.

3. Currently, we upgrade Vitess by rebasing our patches on top of the upgraded source. This PR has a large blast radius, which could make it difficult to resolve merge conflicts and ensure the patched logic is safely applied to any updates to vindex logic in the source. h/t @ellayarmogray for pointing this out.

### 1. Passing in a pointer to created vindex instances in addition to the vschema info 

@jmchen28 This is the initial approach we were aiming for after our pairing session.

When vindexes from the vschema are created, they are stored in the [ksvschema.Vindexes](
https://github.com/etsy/vitess/blob/a9fdeb467d6088c87d52c0d5f251ba7fa138b7be/go/vt/vtgate/vindexes/vschema.go#L266) struct. We could pass this struct, in addition to the vschema info, into NewVindexFunc.

This would allow us to do the following:
- In the NewVindexFunc, validate that all the subvindexes it needs actually exist in the vschema. This way, we can be confident that they will be created eventually. However, do not validate that instances of subvindexes have been created.
- When the Multisharded vindex's methods that rely on subvindex instances are called, access the subvindex instances via `ksvschema.Vindexes`.

I didn't go this route because I wasn't able to validate that all calls to Multisharded methods that rely on subvindex instances (`Map`, `Verify`, `Cost`, `IsUnique`, `NeedsVCursor`) will be called _after_ all vindexes defined in vschema.json are loaded. It felt safer to guarantee access to subvindex instances when a Multisharded vindex is created.

I considered adding logic in each of the Multisharded methods that rely on subvindex instances to check if the instance already existed and to create (and store a pointer to it on the Multisharded struct) the subvindex if the it did not already exist. However, this would require updating the Vindex interface so that `Cost`, `IsUnique`, and `NeedsVCursor` also returned an `error` (since `CreateVindex` may return an error) and handling that error. I didn't feel confident in my familiarity with the Vitess codebase to safely make this update and handle these errors.

### 2. Passing in subvindex configs as a "param" value for the multisharded vindex in vschema.json. 

An example of this would be:
```
        "etsy_multisharded_hybrid": {
            "type": "etsy_multisharded_hybrid",
            "params": {
                "type_id_to_vindex": "{\"2\":\"etsy_shard_hybrid_facebook_account\",\"30\":\"etsy_shard_hybrid_shop\"}",
                "subvindexes": "{\"etsy_shard_hybrid_shop\":\"params\": {\"cache_size\": \"2000\",\"from\": \"shop_id\",\"path\": \"dbindex.db\",\"table\": \"shops_index\",\"threshold\": \"123",\"to\": \"ksid\",\"vindex_a\": \"etsy_sqlite_lookup_unique\",\"vindex_b\": \"hash\"}, \"etsy_shard_hybrid_user\":\"params\": {\"cache_size\": \"2000\",\"from\": \"user_id\",\"path\": \"dbindex.db\",\"table\": \"shops_index\",\"threshold\": \"1123\",\"to\": \"ksid\",\"vindex_a\": \"etsy_sqlite_lookup_unique\",\"vindex_b\": \"hash\"}}"
            }
        }
```

This approach would allow us to avoid passing in an extra argument to the NewVindexFunc, but it would make the vschema.json verbose, especially because the `"params":` key currently only supports string values. And the Multisharded vindex would still create duplicate vindexes.

### 3. When the vschema.json is loaded, sort the vindex configs so they are loaded in a predictable way.

One option is to sort the vindex configs alphabetically in [BuildTables](https://github.com/etsy/vitess/blob/a9fdeb467d6088c87d52c0d5f251ba7fa138b7be/go/vt/vtgate/vindexes/vschema.go#L252), and ensure that Multisharded vindexes are defined alphabetically after any of their subvindexes in vschema.json.

This approach would be viable and low-effort, but it felt a little hacky. We'd need to append something like `z_` to each multisharded vindex name in vschema.json to ensure it comes alphabetically behind its subvindexes.

This might be a good approach if we are limited in time.

### 4. Use [json schema references](https://niem.github.io/json/reference/json-schema/references/) to cross reference the configs for Hybrid vindexes in the configs for Multisharded vindexes
This would make the vschema.json less verbose, but it would require us to update Vitess to be able to parse json references, which seemed more complex than the Vitess updates in the current approach. The Multisharded vindex would still create duplicate vindexes.

We also had some (initial, untested) performance concerns around parsing json schema references upon loading a vschema.json file.

## Testing

- Updated the Multisharded vindex's unit tests
- In etsy-vitess-sandbox, tested loading a vschema with a Multisharded vindex and reading and writing records to a multisharded table.
- Confirmed all unit tests under the `go/vt/vtgate/{engine|planbuilder|semantics|vindexes}` directories pass.  
    - This PR also has minor changes to files under `go/vt/wrangler` and `go/vt/vttablet/tabletserver/vstreamer` but I ran into errors related to mysql infrastructure dependencies not being set up in the testing environment. Based on [this doc](https://docs.google.com/document/d/1Pa7KWxmRtgud9ugt8D0mv5LZUhihOlefvBoMlmzjIuI/edit#heading=h.xgl98b6fomy0), it doesn't look like we have the ability to run all unit test without opening a PR in the planet scale repo.)

I didn't repeat performance testing, which was done when the Multisharded vindex was initially added, since our perf testing approach focused on query routing, and this PR doesn't change much at the query routing phase (just at the phase when the vschema is loaded).

## Related docs
- Brainstorming doc/notes from pairing with @jmchen28 on this specific issue: https://docs.google.com/document/d/1K7cVhGslNdpiViP75uKJ5zw5CIvQg37BFcO9L3KUEs4/edit#heading=h.7ihph5aaersz
- @wsung-asong's initial investigation into creating a multisharded vindex: https://docs.google.com/document/d/1VOi6xEgkeC69soahgdEp7zuC8E64xQSVm_NGTN0MnCk/

### CODE REVIEW

**Comment by ellayarmogray** [comment]
I haven't looked thoroughly through the code yet, but from reading through the PR description --

I am a bit wary of modifying so many existing Vitess files because of the increased drift from the source code. This could make it harder to upgrade and cherry pick patches, especially for those who are less familiar with vindex code, since (as of now) we have to reapply any changes on top of the source when we upgrade. I'd be interested in weighing that with the downsides of options 2 and 3. 

Option 2 seems reasonable to me – there are not that many multisharded models, and I'm curious what the cons are of having a verbose vschema aside from readability, since its only for a small subset of tables. Maybe it could be possible to generate it somehow.

I could also see an argument for 3 – since this is not a large set of tables and they're not created frequently, I could see this making sense a minimal change, especially if we want to add future work to re-architect/re-shard multisharded tables to do something like shard by hash of any id, or split up the multisharded table by shardifier (I feel like this was discussed, but I can't find where)

I am sure there's been a lot of discussion and thought around this that I haven't been involved in, so forgive me if I'm missing something in how this was decided on!

**Comment by dasl-** [comment]
Nice debugging! It's too bad that the initial approach didn't work the way we hoped it would. I share Ella's concerns about this large patch making it harder to upgrade to future vitess versions (more potential for merge conflicts). That might be a good thing to add to the "Risks" section of your PR. If we thought we could get this PR merged upstream, then I'd be more ok with this approach.

IIUC option 2 would mean we need to define `etsy_shard_hybrid_shop` multiple times in our vschema.json? I don't love that...

I agree that option 3 is hacky, but it would work, and it would be a very minimal surface area for merge conflicts when we upgrade in the future. So I actually like that option. 

A variant of option 3 is that we could retry creating whichever vindexes failed due to the ordering issue. I'm not sure if this is any better or worse than sorting the vindexes. Excuse my very poor pseudocode:
```
vindexesToRetry[] 
for vname, vindexInfo := range ks.Vindexes {
    try {
        createVindex(vname, vindexInfo)
    } catch {
        vindexesToRetry[] = vname, vindexInfo
    }
}

for vname, vindexInfo := range vindexesToRetry {
    createVindex(vname, vindexInfo)
}
```

Anyway, TLDR I wonder if option 3 or something similar to it would be easier to maintain.

**Comment by swu-etsy** [comment]
@ellayarmogray @dasl- That's a fair point about the large number of files making Vitess upgrades more toilsome. I may not have given enough weight to this risk initially, not having been through a Vitess upgrade 😅. I was hoping that each file having a small change would mitigate this, but I see the point about triggering a lot of merge conflicts. 

What I initially liked about this PR's approach over Options 2 & 3 is that this PR's approach would both support a large number of and more complicated dependency relationships between vindexes. Under Option 3, this would require somewhat complicated accounting to confirm that each dependent vindex came after its dependency. Under Option 2, more vindexes that depend on subvindexes means more duplicated subvindex configs in the vschema.json. 

The ability to handle more and/or complex dependencies would mean this PR's approach could support new vindex types that also have subvindex dependencies. This feature felt like it could be a valuable contribution to upstream in the future. On the other hand, we are planning to deprecate Multisharded models, and its hard to imagine our team having a concrete use case for a high number of, or more complicated, vindex dependency relationships. 

I won't have bandwidth to upstream this change before our next Vitess upgrade, so I'm happy to go with an alternative that has less of a blast radius if it means less toil with maintaining our Vitess fork. We can always revisit this implementation if we want to contribute it upstream.

Between Options 2 and 3, I have a slight preference for Option 3, but I would be open to either.

My concern about Option 2, in addition to just adding to the vschema length and reducing readability, is that it'll require us to write and maintain more tooling to ensure correct json. As @ellayarmogray suggested, we'd want a script to generate the duplicate subvindex configs, and we'd likely want to update our vschema.json GitHub check to validate that the duplicate configs for the subvindexes are identical to their separately defined counterparts.

But I'm also not opposed to Option 2. It looks like there are only <15 shardifiers currently in use by [multisharded models](https://docs.google.com/spreadsheets/d/1C4wyLuIsqQyb4cz4tp-1IEwwiPU-sgyw42bmjUrilbg/edit#gid=0), which would limit the amount of duplicated json at least.

By contrast, Option 3 would only require us to validate in the vschema.json Github check that the multisharded models are alphabetized last. It'd also have the benefit of not creating duplicate subvindex instances in memory. A possible downside to Option 3 is that if a Vitess upgrade ever added additional/alternative paths in the Vitess repo where vindexes are created from the vschema, we'd need to catch those and add the sorting logic to those sites. (This is also a downside of this current PR.)

I'm also open to the retry route that @dasl- suggested. The nice thing about retrying is that we wouldn't need to add any additional json validation logic. The downside is that it would only support one layer of vindex dependencies. (E.g., it wouldn't support a situation where `VindexA` depended on `VindexB` which depended on `VindexC`, since a vindex that failed to find subvindexes would only be retried once.

Does anyone have strong feelings between Option 2 or either variant of Option 3 (alphabetical loading v. retrying)?

**Comment by adsr** [comment]
I also haven't gone through the code yet but nice hacking @swu-etsy!

> Specifically, the previous approach stored [a pointer to each hybrid vindex](https://github.com/etsy/vitess/blob/a9fdeb467d6088c87d52c0d5f251ba7fa138b7be/go/vt/vtgate/vindexes/hybrid.go#L89) as the vindex was created, and [validated](https://github.com/etsy/vitess/blob/a9fdeb467d6088c87d52c0d5f251ba7fa138b7be/go/vt/vtgate/vindexes/multisharded.go#L64) that all hybrid vindexes that the Multisharded vindex depended on (I'll refer to them as the Multisharded vindex's "subvindexes") existed at the time the Multisharded vindex was created. As it turns out, the order in which vindexes are created is non-deterministic.

I'm probably missing context but could we fetch references to the subvindexes lazily? That would have the downside of not being able to validate in the constructor.

We could potentially add another function to the Vindex interface for initialization to be called after all vindexes have been instantiated.

Sorting sounds fine to me though, and I'm guessing would be fairly uncontroversial to upstream?

**Comment by swu-etsy** [comment]
@adsr Thanks!  Re lazy loading: This was sort of the idea behind Alternative 1 (don't require the subvindexes to exist until `Map`, `Verify`, `Cost`, etc are called). But I ultimately wasn't confident in ruling out the possibility that these vindex methods could be called before all vindexes defined in vschema.json were created. It's possible I'm being overly cautious here.

In terms of adding a method to the Vindex interface that gets called after all vindexes are instantiated --  this could be an elegant solution if we could confidently confirm that no Vindex instance methods get invoked until all Vindexes are instantiated. But adding a method to the Vindex interface would also require touching a lot of files (at least updating all structs that implement the Vindex interface), which raises the same concerns around making it difficult to maintain our fork/apply our patch to upgrades that this PR raises.

Re sorting being uncontroversial to upstream: Hard to say! I could see it being uncontroversial because it's a minimal change, but I could also see it being disfavored as a way to control vindex dependencies on other vindexes because it's not a super complete solution. For example, it doesn't have great support for nested dependencies, and complex dependency relationships would be hard to maintain with just an alphabetical sort of the vschema.json. 😬

**Comment by swu-etsy** [comment]
@tspotts had an interesting idea of adding an optional `sort_order` param to each Vindex definition in the vschema.json.

We could then update the vindex loading logic so that all vindexes with `sort_order: 1` are loaded before all vindexes with `sort_order:2` and so on. And within each sort order cohort, the vindexes would be loaded in a nondeterministic order.

**Comment by swu-etsy** [comment]
As of now, I'm leaning toward @dasl- 's suggestion of catching errors related to missing subvindexes and retrying the dependent vindexes after all other vindexes are instantiated. I like this approach because it hides the implementation detail for loading order from developers who are adding vindexes to vschema.json and cuts down on the amount of json validation we'd need to add to the GitHub check. (This approach would also keep the possibility open of implementing some sort of topological sorting based on a dependency list passed into a vindex's config in the future, if we want to support more complex dependency relationships.) 

Let me know if anyone has concerns about this retry approach! I'd like to get @jmchen28's thoughts when she gets back from PTO Monday, too.

**Comment by swu-etsy** [comment]
Ok! Here's a second crack at solving this bug that uses the retry approach discussed in this thread. https://github.com/etsy/vitess/pull/25

I'll close this PR for now in favor of the new one.
